app/decorators/spree/line_item_decorator.rb

```diff
@@ -4,6 +4,20 @@ def self.prepended(base)
       base.has_many :ad_hoc_option_values_line_items, dependent: :destroy
       base.has_many :ad_hoc_option_values, through: :ad_hoc_option_values_line_items
       base.has_many :product_customizations, dependent: :destroy
+
+      base.validate :product_customization_should_exist_if_required
+    end
+
+    def product_customization_should_exist_if_required
+      existing_product_customization_types = product_customizations.map { |product_customization| product_customization.product_customization_type }.uniq
+      product_customization_types = product.product_customization_types
+      product_customization_types.each do |product_customization_type|
+        if product_customization_type.is_required?
+          unless existing_product_customization_types.include? product_customization_type
+            errors.add(:order, "missing required product customization: #{product_customization_type.name.capitalize}")
+          end
+        end
```

Comment:
The nested conditional can be merged into a single line with the parent conditional:
```suggestion
        if product_customization_type.is_required? && !existing_product_customization_types.include? product_customization_type
          errors.add(:order, "missing required product customization: #{product_customization_type.name.capitalize}")
        end
```